File: app/src/main/java/net/harimurti/tv/extra/AsyncSleep.java
package net.harimurti.tv.extra;

import android.content.Context;
import android.os.Handler;

public class AsyncSleep {
    private Task task = null;
    private Context context;

    public interface Task {
        default void onCountDown(int left){}
        default void onFinish(){}
    }

    public AsyncSleep task(Task task) {
        this.task = task;
        return this;
    }

    public AsyncSleep(Context context) {
        this.context = context;
    }

    public void start(int second) {
        for (int i = 1; i <= second; i++) {
            int left = second - i;
            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    runOnUiThread(()-> {
                        task.onCountDown(left);
                        if (left == 0) {
                            task.onFinish();
                        }
                    });
                }
            }, i * 1000);
        }
    }

    private void runOnUiThread(Runnable task) {
        new Handler(context.getMainLooper()).post(task);
    }
}

File: app/src/main/java/net/harimurti/tv/extra/AsyncSleep.kt
package net.harimurti.tv.extra

import android.os.Handler
import android.os.Looper
import net.harimurti.tv.App

class AsyncSleep {
    private val context = App.context
    private var task: Task? = null
    private var handler: Handler? = null

    interface Task {
        fun onCountDown(count: Int) {}
        fun onFinish() {}
    }

    fun task(task: Task?): AsyncSleep {
        this.task = task
        return this
    }

    fun start(second: Int) {
        handler = Handler(Looper.getMainLooper())
        for (i in 1..second) {
            val left = second - i
            handler?.postDelayed({
                runOnUiThread {
                    task!!.onCountDown(left)
                    if (left == 0) {
                        task!!.onFinish()
                    }
                }
            }, (i * 1000).toLong())
        }
    }

    fun stop() {
        handler?.removeCallbacksAndMessages(null)
        handler = null
    }

    private fun runOnUiThread(task: Runnable) {
        Handler(context.mainLooper).post(task)
    }
}